### PR TITLE
#60704: Pass cert for TLS localhost connection

### LINKED ISSUE #60704: Provide a way to serve localhost through HTTPS
## Use case

I need to test input validations and autofills on a form that takes credit card information.
When running `flutter run -d web-server` or `flutter run -d chrome` it opens up a website served on localhost through HTTP. That disallows the autofill for credit card information as you can see on this screenshot:

Screen Shot 2020-07-01 at 12 26 56 PM

If I were in control of the web-server configuration, I could maybe pass in some options to include an SSL certificate for localhost. Like its mentioned here: https://stackoverflow.com/questions/54735136/how-to-enable-chrome-credit-card-autofill-for-http-and-localhost

## Proposal

Need a way to pass SSL certs (and maybe other options) when starting up a web-server through flutter

### PR DESCRIPTION
**Pull request by arpitgandhi9**
*Pass locally generated certificate via command line*

*Fixes: #60704*

Added ARGS: 

- web-tls-cert-path
- web-tls-cert-key-path

Passing the path of local certificate and the key to cert will allow flutter tool to create a secure debugging session on chrome

**Pre-launch Checklist**

✅   I read the [Contributor Guide](https://github.com/flutter/flutter/wiki/Tree-hygiene#overview) and followed the process outlined there for submitting PRs.
✅   I read the [Tree Hygiene](https://github.com/flutter/flutter/wiki/Tree-hygiene) wiki page, which explains my responsibilities.
✅   I read and followed the [Flutter Style Guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo), including [Features we expect every widget to implement](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement).
✅   I signed the [CLA](https://cla.developers.google.com/).
✅   I listed at least one issue that this PR fixes in the description above.
✅   I updated/added relevant documentation (doc comments with ///).
✅   I added new tests to check the change I am making.
✅   All existing and new tests are passing.

### CODE REVIEW

**Comment by arpitgandhi9** [comment]
could #103230 be the issue for windows misc to fail?

**Comment by c-seeger** [comment]
this would be awesome if we get this merged!

**Comment by christopherfujino** [comment]
I pinged the discord chat about looking at this PR

**Comment by christopherfujino** [comment]
> I pinged the discord chat about looking at this PR

Following up from a discussion thread on the discord, it sounds like this isn't actually needed for most developer use cases. @ditman & @yjbanov shall we close this as won't fix?

**Comment by c-seeger** [comment]
It would be extremely helpful to get this. I'm currently working on an app using SSO provider that require TLS to be present currently I need to proxy the requests to do the same or run `flutter build web` and use a webserver with https to serve these files.

**Comment by christopherfujino** [comment]
I'm sorry, but I am going to be closing this PR as something we are not able to support at this time.

**Comment by jacobwardio** [comment]
@christopherfujino why are you unable to support at this time?

**Comment by c-seeger** [comment]
> > I pinged the discord chat about looking at this PR
> 
> Following up from a discussion thread on the discord, it sounds like this isn't actually needed for most developer use cases. @ditman & @yjbanov shall we close this as won't fix?

@christopherfujino, @ditman, @yjbanov  I'm not sure if this is true here. 

This is a web only feature since you only need it for flutter web development, mobile or desktop does not need this. So when asking flutter developers if this is a needed feature, all that use flutter only for mobile or desktop development will say no. 

For web development it is nessesary since TLS is a requirement for many 3rd party integrations. So the question should go to **flutter web developers** and according to this issue and https://github.com/flutter/flutter/issues/60704 there is enough evidence that this feature is needed. 

As already stated in the mentioned issue, without this feature you have to use flutter in a crippled way when developing since you lose the hot reload as well as device specific debug features that come with `flutter run` which is not usable without providing a way to make TLS work once you hit the TLS barrier of a 3rd party integration. Most commonly SSO providers! Which is a very common use case for web applications. 

Currently flutter is not widely adapted to be used for web applications yet, and in my opinion having no support for TLS with `flutter run` is one big entry barrier here.

**Comment by ditman** [comment]
> TLS is a requirement for many 3rd party integrations.

@c-seeger can you please be more specific about this? "many" sounds to me like a [_weasel word_](https://en.wikipedia.org/wiki/Weasel_word) in this discussion, but **concrete use cases** would go a long way to _sell_ the value of this feature, and re-prioritize its landing.

All browsers now consider `localhost` as a secure context[^1], so what I'm wondering is:

**Is `flutter run` the right layer at which this fix needs to be applied**, or should this be fixed by the 3rd party that requires `https` in `localhost` for their service? (And what's the threat model that inspired that requirement on the 3rd party?)

> the question should go to flutter web developers

The **[question](https://discord.com/channels/608014603317936148/608022171637252115/1005233131378454579) did go** to our Discord channel `#hackers-web` (invite link to the server [here](https://github.com/flutter/flutter/wiki/Chat)). I asked what the use cases for this were, and I didn't get an answer.

It is also unclear that this would fix the original issue (Chrome refusing to auto-fill CC fields), see this [SO answer](https://stackoverflow.com/a/66685223) (and a fix without certs).

[^1]: https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts#when_is_a_context_considered_secure

**Comment by c-seeger** [comment]
Thanks @ditman for your quick reply.

You are right concret use cases are needed:

I'm currently developing a web/mobile application that uses SSO logins (google, facebook, apple). 

The usual workflow here is
- Browser send request to google/facebook/apple
- Browser gets a token back and sends it to our backend
- Our backend validates the SSO token and replies with JWT token (for web in form of http secure only cookie)

Currently if I just use flutter run I only get http://localhost and requesting from http all SSO provider refuse the request due to missing TLS. I also tried the chrome `insecure origins thread as secure` which wasn't working for me either, i still get the same result.

The current workaround I'm using is running a webserver and using `flutter build web` instead of using `flutter run` .

This comes with costs like no hot reload, which slows down implementation speed :(.

> The [question](https://discord.com/channels/608014603317936148/608022171637252115/1005233131378454579) did go to our Discord channel #hackers-web (invite link to the server [here](https://github.com/flutter/flutter/wiki/Chat)). I asked for use cases for this, and didn't get an answer.

I'm sorry I wasn't aware of this but will join the discord as well (thanks for the link).

> It is also unclear that this would fix the original issue (Chrome refusing to auto-fill CC fields), see this [SO answer](https://stackoverflow.com/a/66685223) (and a fix without certs).

Thats indeed a good question since I only run into TLS issues using 3rd party integrations, I can not answer this.

Other examples from the https://github.com/flutter/flutter/issues/60704
- auto-fill CC (as mentioned unclear)
- authentication (google, apple, facebook, ... my guess is this is also true for all other SSO providers like github, etc)
- grpc backend (no additional context mentioned)
- payment service (indeed!)
- firebase features (not exactly mentioned which ones)
 
I hope this helps in clarifying the use cases a bit more. Any solution that helps to make `flutter run` working with the TLS requirement is very appreciated. Thanks again for taking care of this :heart:

**Comment by c-seeger** [comment]
an addition:

> Is flutter run the right layer at which this fix needs to be applied, or should this be fixed by the 3rd party that requires https in localhost for their service? (And what's the threat model that inspired that requirement on the 3rd party?)

for authentication (as well as payment) it makes sense to enforce the origin to have TLS no matter if it's just a test environment or not. The attack vector here would be the possibility of a network intruder to be able to read sensible data via MITM (in our case the initial token returned from the SSO provider, allowing to forge a valid request against our backend to obtain valid JWT tokens).

https://stackoverflow.com/questions/7594193/calling-https-from-http-through-ajax-for-login

**Comment by nicktgn** [comment]
Any plans to get this merged?

**Comment by Schwusch** [comment]
It is a pain to work with APIs like WebUSB/WebSerial/WebNFC or Credential Management API which require HTTPS.
My current workaround is to use [package:flutter_cors](https://pub.dev/packages/flutter_cors) to turn off all security while testing, but that does not help with WebNFC, since it only exists on Android, where I can't turn security off.

**Comment by ditman** [comment]
(I've been convinced that we should land this (or a similar version). It'll happen, especially if anybody other than myself refreshes this PR and brings it to mergeability :) )

**Comment by arpitgandhi9** [comment]
@ditman Merged and pushed. You can reopen this if required?

**Comment by ditman** [comment]
@arpitgandhi9 this is failing CLA with the user @arpitignite (are both accounts yours?)

File: packages/flutter_tools/lib/src/base/io.dart [hunk 1]
```diff
@@ -80,6 +80,7 @@ export 'dart:io'
         HttpStatus,
         IOException,
         IOSink,
+        SecurityContext,
```
Comment (ditman) [inline packages/flutter_tools/lib/src/base/io.dart:83]:
```
Analyzing flutter sdk...                                        

   info • Sort combinator names alphabetically • packages/flutter_tools/lib/src/base/io.dart:63:5 • combinators_ordering

1 issue found. (ran in 294.2s)
```

(This should be moved to line 96, above `ServerSocket`)

File: packages/flutter_tools/test/general.shard/web/devfs_web_test.dart [hunk 2]
```diff
@@ -1134,6 +1148,8 @@ void main() {
     final WebDevFS webDevFS = WebDevFS(
       hostname: 'localhost',
       port: 0,
+      tlsCertPath: null,
+      tlsCertKeyPath: null,
```
Comment (ditman) [inline packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:1152]: Any chance of adding a test where these are not `null`, and we can verify that the server is bound to `https`?

Comment (arpitgandhi9) [reply packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:1152]: This might end up being a little complicated than its purpose since the certificate need to be generated and then passed and then the connection will have to be checked. I can try but it might take sometime since this is not elemetary.

Comment (ditman) [reply packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:1152]: @arpitgandhi9 I'd appreciate it if you gave it a fair shot, that way the feature is harder to "revert". If needed, an empty certificate and key can possibly be added as test assets?

**Comment by christopherfujino** [comment]
@arpitgandhi9 are you still interested in working on this? And if not, @ditman should we close this, or do you want to take it over?

**Comment by ditman** [comment]
@christopherfujino let's keep it open, this has been requested by several people.

File: packages/flutter_tools/test/general.shard/web/devfs_web_test.dart (hunk 2 above)
Comment (arpitignite) [reply packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:1152]: Just checked this, was caught up in some personal commitments. Will check this asap.

Comment (arpitgandhi9) [reply packages/flutter_tools/test/general.shard/web/devfs_web_test.dart:1154]: @ditman Added test case to validate connection URI with a dummy certificate. This certificate is part of a dummy test asset package.

**Comment by arpitgandhi9** [comment]
> @arpitgandhi9 are you still interested in working on this? And if not, @ditman should we close this, or do you want to take it over?

Done. Was caught up in some personal commitments.

**Comment by arpitgandhi9** [comment]
@ditman Can you please help with why this is failing? It is working fine locally.

**Comment by M144-Coder** [comment]
Will this feature get merged any time soon? We need this feature for local development and testing with services that require secure origins. Thanks for the great efforts.

**Comment by christopherfujino** [comment]
> Will this feature get merged any time soon? We need this feature for local development and testing with services that require secure origins. Thanks for the great efforts.

Helping the author investigate test failures would expedite approving and merging this change.

File: packages/flutter_tools/test/data/asset_test/main/pubspec.yaml [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+description: A test project that has font and dummy certificates packages as dependencies.
 
 environment:
   sdk: '>=3.0.0-0 <4.0.0'
 
 dependencies:
   font:
     path: ../font
+  tls_cert:
+    path: ../tls_cert
```
Comment (ditman) [inline packages/flutter_tools/test/data/asset_test/main/pubspec.yaml:11]: I think the test is failing because we're defining tls_cert as a package with assets, and that's interfering with what's expected in the other test.

I think the fix to this is to make the `tls_cert/dummy-cert.pem` and `tls_cert/dummy-key.pem` just test fixtures (files in the disk without any special flutter "meaning"), and work to figure out what the correct path to them would be from the test where they're tested!

Comment (arpitgandhi9) [reply packages/flutter_tools/test/data/asset_test/main/pubspec.yaml:11]: Thanks for reviewing. Let me try to place this somewhere else.

Comment (arpitgandhi9) [reply packages/flutter_tools/test/data/asset_test/main/pubspec.yaml]: @ditman Done!! Good to go.

**Comment by yjbanov** [comment]
@ditman Is this good to go?

**Comment by christopherfujino** [comment]
friendly ping @ditman

**Comment by arpitgandhi9** [comment]
@ditman Anything else needed here?

**Comment by christopherfujino** [comment]
friendly ping @ditman

**Comment by christopherfujino** [comment]
friendly ping @ditman

**Comment by holocronweaver** [comment]
Paging Dr. @ditman , paging Dr. @ditman

**Comment by christopherfujino** [comment]
@ditman friendly ping

**Comment by christopherfujino** [comment]
pinged the web channel on discord

File: packages/flutter_tools/lib/src/isolated/devfs_web.dart [hunk 4]
```diff
@@ -782,10 +797,13 @@ class WebDevFS implements DevFS {
...
     }
+    String url = '$hostname:$selectedPort';
     if (hostname == 'any') {
-      _baseUri = Uri.http('localhost:$selectedPort', webAssetServer.basePath);
-    } else {
-      _baseUri = Uri.http('$hostname:$selectedPort', webAssetServer.basePath);
+      url ='localhost:$selectedPort';
+    }
+    _baseUri = Uri.http(url, webAssetServer.basePath);
+    if (tlsCertPath !=null && tlsCertKeyPath!=null) {
```
Comment (ditman) [inline packages/flutter_tools/lib/src/isolated/devfs_web.dart:805]:
```suggestion
    if (tlsCertPath != null && tlsCertKeyPath!= null) {
```

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart [hunk 5]
```diff
@@ -335,7 +337,11 @@ Please provide a valid TCP port (an integer between 0 and 65535, inclusive).
           mainPath: target,
           debuggingOptions: debuggingOptions,
           platformArgs: <String, Object>{
-            'uri': url.toString(),
+             'uri': (debuggingOptions.tlsCertPath != null &&
+                         debuggingOptions.tlsCertKeyPath != null
+                     ? url.replace(scheme: 'https')
+                     : url)
+                 .toString(),
```
Comment (ditman) [inline packages/flutter_tools/lib/src/isolated/resident_web_runner.dart:344]: Please, move this manipulation of the `url` to right next where it's created (Line ~305).

Also maybe check that both tlsCertPath and tlsCertKeyPath are different than null (or equal null) at the same time?

**Review by ditman — APPROVED** [review]
We need a blog post/page in the doc site explaining why this was needed, and how to use it when developing locally IMO!

**Comment by eliasyishak** [comment]
@arpitgandhi9 friendly reminder from our PR review for the requested changes from @ditman. Thank you for contribution!

File: packages/flutter_tools/lib/src/isolated/resident_web_runner.dart (hunk 5 above)
Comment (arpitgandhi9) [reply packages/flutter_tools/lib/src/isolated/resident_web_runner.dart]: Something like this?

**Comment by arpitgandhi9** [comment]
@ditman Changes done. Please check

**Review by christopherfujino — APPROVED** [review]
RSLGTM

**Comment by jwalton** [comment]
I spent a while banging my head against a wall trying to get nginx to forward to flutter, before I discovered this PR has already fixed this.  Could some admin (@ditman @christopherfujino ??) please go add a comment to the bottom of https://github.com/flutter/flutter/issues/60704 to let people know this PR has been merged and this is fixed in master?  Many thanks!

**Comment by ditman** [comment]
Good call @jwalton, I just added a comment at the end of the locked issue. Thanks for flagging this!
